module/VuFind/src/VuFind/RateLimiter/RateLimiterManagerFactory.php

```diff
@@ -0,0 +1,176 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * @category VuFind
+ * @package  Service
+ * @author   Ere Maijala <ere.maijala@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development Wiki
+ */
+
+namespace VuFind\RateLimiter;
+
+use Closure;
+use Laminas\Cache\Psr\CacheItemPool\CacheItemPoolDecorator;
+use Laminas\ServiceManager\Exception\ServiceNotCreatedException;
+use Laminas\ServiceManager\Exception\ServiceNotFoundException;
+use Laminas\ServiceManager\Factory\FactoryInterface;
+use Psr\Container\ContainerExceptionInterface as ContainerException;
+use Psr\Container\ContainerInterface;
+use Symfony\Component\RateLimiter\LimiterInterface;
+use Symfony\Component\RateLimiter\RateLimiterFactory;
+use Symfony\Component\RateLimiter\Storage\CacheStorage;
+use Symfony\Component\RateLimiter\Storage\StorageInterface;
+use VuFind\RateLimiter\Storage\CredisStorage;
+
+/**
+ * Rate limiter manager factory.
+ *
+ * @category VuFind
+ * @package  Service
+ * @author   Ere Maijala <ere.maijala@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development Wiki
+ */
+class RateLimiterManagerFactory implements FactoryInterface
+{
+    /**
+     * Service locator
+     *
+     * @var ContainerInterface
+     */
+    protected $serviceLocator;
+
+    /**
+     * Create an object
+     *
+     * @param ContainerInterface $container     Service manager
+     * @param string             $requestedName Service being created
+     * @param null|array         $options       Extra options (optional)
+     *
+     * @return object
+     *
+     * @throws ServiceNotFoundException if unable to resolve the service.
+     * @throws ServiceNotCreatedException if an exception is raised when
+     * creating a service.
+     * @throws ContainerException&\Throwable if any other error occurs
+     */
+    public function __invoke(
+        ContainerInterface $container,
+        $requestedName,
+        array $options = null
+    ) {
+        if (!empty($options)) {
+            throw new \Exception('Unexpected options sent to factory.');
+        }
+
+        $this->serviceLocator = $container;
+
+        $yamlReader = $container->get(\VuFind\Config\YamlReader::class);
+        $config = $yamlReader->get('RateLimiter.yaml');
+
+        $authManager = $container->get(\VuFind\Auth\Manager::class);
+        if ($user = $authManager->getUserObject()) {
+            $id = 'u:' . $user->getId();
```

Comment:
Might there be a reason to make it a config setting to account for users? Might there be use cases where we always want to filter by IP address? (I'm thinking, for example, of an attack where somebody automatically creates lots of accounts... that would effectively bypass the rate limiter).

Reply:
I've now added the `preferIPAddress` setting to policies. This along with the new `ipRanges` filter should improve the configurability.

Reply:
But I just realized that it doesn't make sense to `ipRanges` should probably be higher level, i.e. not filter-specific. Will revise..

Reply:
Okay, should be good now. I also added `loggedIn` that can be used to limit the policy for logged-in or anonymous clients.